### ISSUE TITLE
rpt2 does not resolve pnpm symlinked dependencies

### ISSUE THREAD

**Issue by davelsan** [ezolenko/rollup-plugin-typescript2#234]
#### What happens and why it is wrong

When trying to build a project installed with [`pnpm`](https://pnpm.js.org/), `rpt2` fails to resolve the created symlinked dependencies in `node_modules` (non-flat).

Setting `check: false` is (arguably) a workaround.

Related issues: #188, #214

#### Reproduction

Repository: https://github.com/davelsan/rpt2-pnpm-symlinks
Config: https://github.com/davelsan/rpt2-pnpm-symlinks/blob/master/rollup.config.js

The provided repro is a minimal `vue3` plugin that imports the `App` interface from `vue` in [`index.ts`](https://github.com/davelsan/rpt2-pnpm-symlinks/blob/7345a15cb9b756a6216fa5b3a0772a7a87314a27/src/index.ts#L1)

```ts
import  { App } from 'vue'
```

This import causes the following error on build:

```
 [!] (plugin rpt2) Error: semantic error TS2305: Module '"../node_modules/vue/dist/vue"' has no exported member 'App'.
```

Setting the `check` option to `false` in [`rollup.config.js`](https://github.com/davelsan/rpt2-pnpm-symlinks/blob/7345a15cb9b756a6216fa5b3a0772a7a87314a27/rollup.config.js#L13) eliminates the error.

#### Versions
- typescript: 4.2.3
- rollup: 2.41.5
- rollup-plugin-typescript2: 0.30.0

#### rollup.config.js

```js
import path    from 'path';
import rpt2    from 'rollup-plugin-typescript2';
import pkgJson from './package.json';

export default {

  input: './src/index.ts',

  external: ['vue'],

  plugins: [
    rpt2({
      check: true,
      tsconfig: path.resolve(__dirname, 'tsconfig.json'),
    })
  ],

  output: [
    {
      file: pkgJson.browser, // dist/index.esm.js
      format: 'es',
      sourcemap: true,
    },
  ],
}
```

#### tsconfig.json

A snippet with the relevant module resolution options only.

```json
{
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
  }
}
```

#### package.json

A snippet with the relevant dependencies only.

```json
{
  "peerDependencies": {
    "vue": "^3.0.7"
  },
  "devDependencies": {
    "rollup": "^2.41.5",
    "rollup-plugin-typescript2": "^0.30.0",
    "typescript": "^4.2.3",
    "vue": "^3.0.7"
  }
}
```

**Comment by davelsan** [comment]
#### Update

According to [this comment](https://github.com/pnpm/pnpm/issues/2165#issuecomment-555752571) in the `pnpm` repo, it could be a problem caused by the `rollup-plugin-commonjs` plugin (note that I'm not using it in the repro).

However, none of the provided solutions will work in this case.

```js
// include the actual pnpm registry path
commonjs({ include: '../../node_modules/.pnpm/registry.npmjs.org/**' })

// or set the following option in rollup.config.js
preserveSymlinks: true
```

**Comment by davelsan** [comment]
I've been looking at the code and I'm mystified. I'm leaving more info here for myself and other passing-by folks:

- In theory `resolve` already supports symlinks (browserify/resolve#196 at was fixed in browserify/resolve#197)

- This is also be supported in `@rollup/plugin-resolve` via the rollup option `preserveSymlinks`. Though it seems to use a different resolve algorithm.
https://github.com/rollup/plugins/blob/9ad41a0760b4eeed9fd26d2da9b5122aa8a351e5/packages/node-resolve/src/index.js#L148-L156

- This plugin does use `resolve.sync` if `rollupCommonJSResolveHack` is set to `true`. But it does not pass an options object.

https://github.com/ezolenko/rollup-plugin-typescript2/blob/428140bd2260e3ff29b1255029331e2c5b436fa2/src/index.ts#L160-L162

Based on the above, I cannot see any combination of options and/or plugins that will correctly resolve `pnpm` symlinks. Maybe I could try to implement it? It looks simple enough, which probably means it is more complicated than I can see 😅

**Comment by davelsan** [comment]
I'm starting to suspect this is not an issue with rpt2, despite what the error says, but rather a problem that's propagating upstream.

Closing until I have more information.

**Comment by creepinson** [comment]
Any updates on this? I'm trying to bundle a react app with rollup and the original @rollup/plugin-typescript gives debug failure errors, so I tried this.

I got the following error specifically:

```
[!] (plugin rpt2) Error: /media/theo/TheoData/Essentials/Projects/Programming/hypervpn/src/pages/home.tsx(1,10): semantic error TS2305: Module '"@chakra-ui/react"' has no exported member 'Button'.
src/pages/home.tsx
Error: /media/theo/TheoData/Essentials/Projects/Programming/hypervpn/src/pages/home.tsx(1,10): semantic error TS2305: Module '"@chakra-ui/react"' has no exported member 'Button'.
    at error (/media/theo/TheoData/Essentials/Projects/Programming/hypervpn/node_modules/.pnpm/rollup@2.40.0/node_modules/rollup/dist/shared/rollup.js:5279:30)
```

**Comment by creepinson** [comment]
Thing is, it works fine with the original ts plugin before I had debug failure errors, and running tsc by itself works fine as well.

**Comment by beetaa** [comment]
> Thing is, it works fine with the original ts plugin before I had debug failure errors, and running tsc by itself works fine as well.

I am using pnmp as the package manager, same situation while rolling up a vue component. no problem when using  `@rollup/plugin-typescript`.

![image](https://user-images.githubusercontent.com/5688224/111403551-1cf15c00-8708-11eb-86b9-7babce819877.png)

**Comment by beetaa** [comment]
@davelsan @creepinson I think this issue should be opened again for further discussion. thank you.

**Comment by davelsan** [comment]
Hi @beetaa. Thanks for your interest in this issue. I have updated the original repro and I can confirm the issue is still there. I'm re-opening the discussion as per your request.

**Comment by ezolenko** [comment]
@davelsan 
> This plugin does use `resolve.sync` if `rollupCommonJSResolveHack` is set to `true`. But it does not pass an options object.
> [...]
> Based on the above, I cannot see any combination of options and/or plugins that will correctly resolve `pnpm` symlinks. Maybe I could try to implement it? It looks simple enough, which probably means it is more complicated than I can see 😅

So you are saying rpt2 should pass `{ preserveSymlinks: false }` option to resolve?

**Comment by jackson-yyy** [comment]
> @davelsan
> 
> > This plugin does use `resolve.sync` if `rollupCommonJSResolveHack` is set to `true`. But it does not pass an options object.
> > [...]
> > Based on the above, I cannot see any combination of options and/or plugins that will correctly resolve `pnpm` symlinks. Maybe I could try to implement it? It looks simple enough, which probably means it is more complicated than I can see 😅
> 
> So you are saying rpt2 should pass `{ preserveSymlinks: false }` option to resolve?

I got the same problem in my project with pnpm.
I debugged line by line and found the problem is `getAllReferences` function returned a path  the symlink, not the path of the real file or directory.
https://github.com/ezolenko/rollup-plugin-typescript2/blob/428140bd2260e3ff29b1255029331e2c5b436fa2/src/tscache.ts#L63-L75

for examples,
consider this following directory
```
node_modules
├─.pnpm
|  ├─vue@3.2.2
|  |  ├─node_modules
|  |  |   ├─@vue
|  |  |   ├─vue
packages
|  ├─demo
|  |   ├─node_modules
|  |   │  └─vue(symlink, real path should be `project_root/node_modules/.pnpm/vue@3.2.2/node_modules/vue`)
|  |   ├─public
|  |   └─src

```

As you see, the real path of vue in demo should be `project_root/node_modules/.pnpm/vue@3.2.2/node_modules/vue`
But after resolving by `tsModule.nodeModuleNameResolver` in function `getAllReferences` , it becomes `project_root/packages/demo/node_modules/vue`

`Vue@3.2.2` depends on `@vue`, which is not exist in `project_root/packages/demo/node_modules`, so it throw an error while resolving.

I wonder if there is any way to solve this problem.
Otherwise i can only use `pnpm --shamefully-hoist` to hoist all packages to root node_modules, which is similar to yarn/npm.😅

**Comment by mammadataei** [comment]
Any news?
I have the same issue 
```
semantic error TS2305: Module '"vue"' has no exported member 'defineComponent'.
```

**Comment by RookieChen4** [comment]
same error.

**Comment by RookieChen4** [comment]
@rollup/plugin-typescript solve this.

**Comment by l3szcz** [comment]
[rollup-plugin-ts](https://www.npmjs.com/package/rollup-plugin-ts) works too and also exports compiled d.ts typings without much of a hassle.

**Comment by agilgur5** [comment]
## Fixed the issue!

So I believe I've found a fix for this long-standing issue! See #332

## Notes on root cause analysis / debugging

### Red herring to look at `resolveId` hook etc

> I've been looking at the code and I'm mystified. Leaving more info here for myself and other passing-by folks:

It was helpful to see this previous research, but this seemed to be a bit of a red herring... I'll explain shortly.

This issue doesn't list "plugin output with `verbosity: 3`" as requested in the issue template, but #330 did, and I confirmed the fix in its reproduction. I also took the reproduction here and added `verbosity: 3` and confirmed it had very similar logs. I also confirmed (by locally editing rpt2 in `node_modules`) that the fix works for this reproduction as well.

Notably, the logs don't output anything about ["resolving"](https://github.com/ezolenko/rollup-plugin-typescript2/blob/e8240ae505a6c4d0b7faa997e3c2d75d293c594a/src/index.ts#L170), so the bug here isn't in the `resolveId` hook _at all_. Which is why looking at `resolve` etc was a red herring.
That portion of the code also resolves directly with TS and with [`tsModule.sys`](https://github.com/ezolenko/rollup-plugin-typescript2/blob/e8240ae505a6c4d0b7faa997e3c2d75d293c594a/src/index.ts#L156) as its host. As I found out later, `ts.sys` has a default implementation of [`realpath`](https://github.com/microsoft/TypeScript/blob/ab2523bbe0352d4486f67b73473d2143ad64d03d/src/compiler/sys.ts#L1288), which turns out to be the important missing piece here.

### Error is in `transform` hook actually

It actually errors right after the ["transpiling"](https://github.com/ezolenko/rollup-plugin-typescript2/blob/e8240ae505a6c4d0b7faa997e3c2d75d293c594a/src/tscache.ts#L212) log statement, which means the error is actually in the `transform` hook, specifically in [`getEmitOutput`](https://github.com/ezolenko/rollup-plugin-typescript2/blob/e8240ae505a6c4d0b7faa997e3c2d75d293c594a/src/index.ts#L203), which is implemented by TS and doesn't have host-specific code for it... But apparently it seems to rely [`host.realpath`](https://github.com/microsoft/TypeScript/blob/67673f324dd5f9398bb53fd16bf75efd155c32e7/src/compiler/moduleNameResolver.ts#L1412) under-the-hood.

### Implementing `host.realpath` fixes this

So I tried implementing `host.realpath` (as #332 shows), and then everything worked and the type-error disappeared!
The typings were generated correctly as well now (#330 had broken typings when using `check: false`).

### TS docs on Compiler APIs suck... 😕  😞 

The TS docs are _incredibly_ sparse when it comes to Compiler integrations such as this plugin. The Compiler API basically has [two](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API) [small](https://github.com/microsoft/TypeScript/wiki/Using-the-Language-Service-API) docs/examples in the TS Wiki... and that's it. Notably, there is no mention of `realpath` what-so-ever in either of those.

Honestly the root cause of a lot of TS integration issues can come down to poor documentation of the Compiler API, which has been noted in _lots_ of places.
As a result, this requires _even more_ significant domain knowledge to figure out issues like this, as the docs aren't even helpful -- I had to literally read through the Compiler source code... and have had to do that a handful of times in the past too. That's obviously not ideal, though at least we can do something like that in open-source.

Basically you need to be a TS Compiler subject matter expert to figure out some of these things, and I'm not sure I would even call myself that despite being a regular contributor to the TS tooling ecosystem (compilers are huge after all!).

### Other Rollup TS plugins

> [@rollup/plugin-typescript](https://www.npmjs.com/package/@rollup/plugin-typescript) solve this.

> [rollup-plugin-ts](https://www.npmjs.com/package/rollup-plugin-ts) works too and also exports compiled d.ts typings without much of a hassle.

I'm honestly not sure how these plugins manage to get around this issue as _neither_ of them implement `host.realpath`

I've mentioned this [elsewhere](https://github.com/ezolenko/rollup-plugin-typescript2/issues/283#issuecomment-1107888296), but why I decided to contribute here (and now I help maintain this repo) as opposed to `rollup-plugin-ts` is because this repo was significantly simpler to wrap my head around. `rollup-plugin-ts` has quite a lot of features -- which is great! But it also means that actually means one needs _even more_ domain knowledge of its structure to contribute.

`@rollup/plugin-typescript` historically wasn't well maintained (that's why this fork came into existence after all), but is quite a bit better maintained nowadays. I believe all 3 plugins (including this one) have different approaches toward various problems, so unfortunately the code can differ quite a bit.
I'd love for TS tooling to use more shared packages though where the approach is the same! E.g. the `host` implementation and `tsconfig` parsing are two areas that could probably be de-duped and am personally working on this in my very limited free time. This is especially useful for really complicated (and poorly documented) subareas of TS.

## Misc

> So you are saying rpt2 should pass `{ preserveSymlinks: false }` option to resolve?

`false` is actually the default of [`resolve`](https://github.com/browserify/resolve#resolveid-opts-cb), [Rollup itself](https://rollupjs.org/guide/en/#preservesymlinks), [TS itself](https://www.typescriptlang.org/tsconfig#preserveSymlinks), which all reflect how Node itself treats symlinks (i.e. `require.resolve`).
And that was a red herring anyway

I also think we may not need `resolve` as a dep at all since we can use built-in Node methods (`resolve` was originally created for `browserify` after all).

**Comment by agilgur5** [comment]
> ### Red herring to look at `resolveId` hook etc

Relatedly (but different root cause), I investigated #295 more after solving this issue and stepped through the resolution logic therein.
I confirmed my suspicion that `resolveId` works correctly due to its direct usage of `tsModule.sys` in  https://github.com/ezolenko/rollup-plugin-typescript2/issues/295#issuecomment-1139678918, where the `"resolving"` logs are output with the correctly resolved `realpath` instead of the one to the symlink.

So not just a red herring, the `resolveId` code actually _correctly_ handles symlinks already.
